### ISSUE TITLE
GATs plus mutable reference causes rustc to improperly believe data escapes function

### ISSUE THREAD

**Issue by Aiden2207** [rust-lang/rust#100135]
I tried this code:

```rust
#![feature(generic_associated_types)]

trait LendingIterator: Sized {
    type Item<'a>
    where
        Self: 'a;
    fn next(&mut self) -> Self::Item<'_>;
}
fn works<I: LendingIterator, F>(iter: I, f: F) -> bool
where
    F: FnMut(I::Item<'_>),
{
    let mut iter2 = Eat(iter, f);
    let _next = iter2.any(|_| false);
    true
}
fn fails<I: LendingIterator, F>(iter: &mut I, f: F) -> bool
where
    F: FnMut(I::Item<'_>),
{
    let mut iter2 = Eat(iter, f);
    let _next = iter2.next();
    true
}
impl<I: LendingIterator> LendingIterator for &mut I {
    type Item<'a> = I::Item<'a> where Self:'a;
    fn next(&mut self) -> Self::Item<'_> {
        (**self).next()
    }
}

struct Eat<I, F>(I, F);
impl<I: LendingIterator, F> Iterator for Eat<I, F>
where
    F: FnMut(I::Item<'_>),
{
    type Item = ();
    fn next(&mut self) -> Option<Self::Item> {
        None
    }
}

```
[Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=05ae5d1cfb5f8f2263aeb25e49864c43)

I expected to see this happen: code successfully compiles.

Instead, this happened: 
```
error[[E0521]](https://doc.rust-lang.org/nightly/error-index.html#E0521): borrowed data escapes outside of function
  --> src/lib.rs:22:17
   |
17 | fn fails<I: LendingIterator, F>(iter: &mut I, f: F) -> bool
   |                                 ----  - let's call the lifetime of this reference `'1`
   |                                 |
   |                                 `iter` is a reference that is only valid in the function body
...
22 |     let _next = iter2.next();
   |                 ^^^^^^^^^^^^
   |                 |
   |                 `iter` escapes the function body here
   |                 argument requires that `'1` must outlive `'static`
   |
   = note: requirement occurs because of a mutable reference to `Eat<&mut I, F>`
   = note: mutable references are invariant over their type parameter
   = help: see <https://doc.rust-lang.org/nomicon/subtyping.html> for more information about variance

For more information about this error, try `rustc --explain E0521`.
error: could not compile `playground` due to previous error
```
Note that removing the bound on `F` causes a successful compilation. Also,  removing the call to `next` lets things compile. The example probably can be minimized more.
### Meta

Playground version:
```
1.64.0-nightly

(2022-08-03 1b57946a405d5b2a87e6
```

**Comment by aliemjay** [comment]
tl;dr `LendingIterator` is pretty much not usable for now unless you force the `'static` bound on the trait.

This is the classic limitation of not supporting implied bounds under `for<'a>` binder. Minimized to generate the same error message:
```rust
#![feature(generic_associated_types)]

trait LendingIterator {
    type Item<'a>
    where
        Self: 'a;
}

impl LendingIterator for &str {
    type Item<'a> = () where Self:'a;
}

fn trivial_bound<I>(_: I)
where
    I: LendingIterator,
    for<'a> I::Item<'a>: Sized,
{
}

fn fails(iter: &str) {
    trivial_bound(iter);
}
```

@rustbot label F-generic_associated_types  A-lifetimes
